src/RandMsg.chpl

```diff
@@ -297,6 +297,64 @@ module RandMsg
         return new MsgTuple(repMsg, MsgType.NORMAL);
     }
 
+    proc segmentedSampleMsg(cmd: string, msgArgs: borrowed MessageArgs, st: borrowed SymTab): MsgTuple throws {
+        const pn = Reflection.getRoutineName(),
+              genName = msgArgs.getValueOf("genName"),                          // generator name
+              permName = msgArgs.getValueOf("perm"),                            // values array name
+              segsName = msgArgs.getValueOf("segs"),                            // segments array name
+              segLensName = msgArgs.getValueOf("segLens"),                      // segment lengths array name
+              weightsName = msgArgs.getValueOf("weights"),                      // permuted weights array name
+              numSamplesName = msgArgs.getValueOf("numSamples"),                // number of samples per segment array name
+              replace = msgArgs.get("replace").getBoolValue(),                  // sample with replacement
+              hasWeights = msgArgs.get("hasWeights").getBoolValue(),            // flag indicating whether weighted sample
+              hasSeed = msgArgs.get("hasSeed").getBoolValue(),                  // flag indicating if generator is seeded
+              seed = if hasSeed then msgArgs.get("seed").getIntValue() else -1, // value of seed if present
+              state = msgArgs.get("state").getIntValue(),                       // rng state
+              rname = st.nextName();
+
+        randLogger.debug(getModuleName(),pn,getLineNumber(),
+                         "genName: %? permName %? segsName: %? weightsName: %? numSamplesName %? replace %i hasWeights %i state %i rname %?"
+                         .doFormat(genName, permName, segsName, weightsName, numSamplesName, replace, hasWeights, state, rname));
+
+        st.checkTable(permName);
+        st.checkTable(segsName);
+        st.checkTable(segLensName);
+        st.checkTable(numSamplesName);
+        const permutation = toSymEntry(getGenericTypedArrayEntry(permName, st),int).a;
+        const segments = toSymEntry(getGenericTypedArrayEntry(segsName, st),int).a;
+        const segLens = toSymEntry(getGenericTypedArrayEntry(segLensName, st),int).a;
+        const numSamples = toSymEntry(getGenericTypedArrayEntry(numSamplesName, st),int).a;
+
+        const sampleOffset = (+ scan numSamples) - numSamples;
+        var sampledPerm: [makeDistDom(+ reduce numSamples)] int;
+
+        if hasWeights {
+            st.checkTable(weightsName);
+            const weights = toSymEntry(getGenericTypedArrayEntry(weightsName, st),real).a;
+
+            forall (segOff, segLen, sampleOff, numSample) in zip(segments, segLens, sampleOffset, numSamples)
+                                                 with (var rs = if hasSeed then new randomStream(real, seed) else new randomStream(real)) {
```

Comment:
yah! so defining it in the `with` means that each task will get it's own private copy of the `randomStream` (these are called task private variables). My primary reason for doing this is because I saw that the `randomStream` records are [not thread safe](https://chapel-lang.org/docs/modules/standard/Random.html#Random.randomStream). So I wanted to avoid any potential issues with that, by having task private variables

Reply:
I have more questions but I'll ask you when I see you in the office.  Thanks!